File: DoiTuyen/Practice/Ksat tháng 4/Test 20.04.2024/PHNT/phnt.py
fi = open("phnt.inp","r")
n = int(fi.readline())
s = fi.readline()
fi.close()

def check(k):
    dic = {}
    for i in range(k,n+1):
        x = s[i-k:i]
        if (dic.get(x) != None):
            return False
        dic[x] = 1
    return True

ans = 0
l, r = 1, n
while (l <= r):
    mid = (l + r) // 2
    if (check(mid)):
        ans = mid
        r = mid - 1
    else:
        l = mid + 1

fo = open("phnt.out","w")
fo.write(str(ans))
fo.close()

File: DoiTuyen/Practice/Ksat tháng 4/Test 20.04.2024/PHNT/phnt.cpp
#include <bits/stdc++.h>

using namespace std;
int n, ans;
string s;
bool check(int k)
{
    set <string> Set; ///cau truc du lieu
    for (int i=k; i<=n; i++)
    {
        string x = s.substr(i-k,k);
        if (Set.find(x) != Set.end())
            return false;
        Set.insert(x);
    }
    return true;
}
int main()
{
    freopen("phnt.inp","r",stdin);
    freopen("phnt.out","w",stdout);
    cin >> n;
    cin >> s;
    int l = 1, r = n;
    while (l <= r)
    {
        int mid = (l + r) / 2;
        if (check(mid))
        {
            ans = mid;
            r = mid - 1;
        }
        else l = mid + 1;
    }
    cout << ans;
    return 0;
}
